code-gen/src/main/scala/scalapb/grpc_web/GrpcServiceMetadataPrinter.scala

```diff
@@ -0,0 +1,277 @@
+package scalapb.grpc_web
+
+import com.google.protobuf.Descriptors.{MethodDescriptor, ServiceDescriptor}
+import scalapb.compiler.FunctionalPrinter.PrinterEndo
+import scalapb.compiler.ProtobufGenerator.asScalaDocBlock
+import scalapb.compiler._
+import scalapb.grpc_web.compat.JavaConverters._
+
+
+final class GrpcServiceMetadataPrinter(service: ServiceDescriptor, implicits: DescriptorImplicits) {
+  import implicits._
+  private[this] def observer(typeParam: String): String = s"$streamObserver[$typeParam]"
+
+  private[this] def serviceMethodSignature(method: MethodDescriptor, overrideSig: Boolean) = {
+    val overrideStr = if (overrideSig) "override " else ""
+     method.streamType match {
+      case StreamType.Unary =>
+        s"${method.deprecatedAnnotation}${overrideStr}def ${method.name}" + s"(request: ${method.inputType.scalaType}, metadata: $metadata): scala.concurrent.Future[${method.outputType.scalaType}]"
+      case StreamType.ServerStreaming =>
+        s"${method.deprecatedAnnotation}${overrideStr}def ${method.name}" + s"(request: ${method.inputType.scalaType}, metadata: $metadata, responseObserver: ${observer(method.outputType.scalaType)}): Unit"
+      case _ =>
+        ""
+    }
+  }
+
+  private[this] def blockingMethodSignature(method: MethodDescriptor, overrideSig: Boolean) = {
+    val overrideStr = if (overrideSig) "override " else ""
+    s"${method.deprecatedAnnotation}${overrideStr}def ${method.name}" + (method.streamType match {
+      case StreamType.Unary =>
+        s"(request: ${method.inputType.scalaType}, metadata: $metadata): ${method.outputType.scalaType}"
+      case StreamType.ServerStreaming =>
+        s"(request: ${method.inputType.scalaType}, metadata: $metadata): scala.collection.Iterator[${method.outputType.scalaType}]"
+      case _ => throw new IllegalArgumentException("Invalid method type.")
+    })
+  }
+
+  private[this] def serviceTrait: PrinterEndo = { p =>
+    p.call(generateScalaDoc(service))
+      .add(s"trait ${service.name} {")
+      .indent
+      .print(service.methods) {
+        case (p, method) =>
+          p.call(generateScalaDoc(method)).add(serviceMethodSignature(method, overrideSig = false))
+      }
+      .outdent
+      .add("}")
+  }
+
+
+  private[this] def blockingClientTrait: PrinterEndo = { p =>
```

Comment:
is blocking client trait needed for grpc-web?

Reply:
will address this in next commit, yes you are right that it's not needed